src/Http/Authenticator.php

```diff
@@ -4,29 +4,208 @@
+     */
+    private $client;
+
+    /**
+     * @var Auth
+     */
+    private $auth;
+
+    /**
+     * @var string
+     */
+    private $accessToken;
+
     /**
-     * @var AuthenticationPlugin
+     * @var string
      */
-    private $authenticationPlugin;
+    private $refreshToken;
+
+    /**
+     * @var int
+     */
+    private $ttl;
+
+    /**
+     * @param Client $client
+     * @param Auth   $auth
+     */
+    public function __construct(Client $client, Auth $auth = null)
+    {
+        $this->client = $client;
+        $this->auth = $auth ?? new NullCredentials();
+    }
 
     /**
-     * @param AuthenticationPlugin $authenticationPlugin
+     * @return array
      */
-    public function __construct(AuthenticationPlugin $authenticationPlugin)
+    public function getTokens(): array
     {
-        $this->authenticationPlugin = $authenticationPlugin;
+        return [
+            'refresh_token' => $this->refreshToken,
+            'token_type' => 'Bearer',
+            'access_token' => $this->accessToken,
+            'expires_in' => $this->ttl,
+        ];
     }
 
     /**
      * @param string $accessToken
+     *
+     * @return Authenticator
+     */
+    public function setAccessToken(string $accessToken): Authenticator
+    {
+        $this->accessToken = $accessToken;
+
+        return $this;
+    }
+
+    /**
+     * @param Client $client
+     *
+     * @return Authenticator
+     */
+    public function setClient(Client $client): Authenticator
+    {
+        $this->client = $client;
+
+        return $this;
+    }
+
+    /**
+     * @return array
+     */
+    public function getAuthHeader(): array
+    {
+        if ($this->accessToken === null) {
+            $this->fetchTokens();
+        }
+
+        return [
+            'Authorization' => 'Bearer '.$this->accessToken,
+        ];
+    }
+
+    /**
+     * @return Auth
      */
-    public function setAccessToken(string $accessToken)
+    public function getAuthCredentials(): Auth
     {
-        $this->authenticationPlugin->setAuthentication(new Bearer($accessToken));
+        return $this->auth;
+    }
+
+    /**
+     * @param string $appId
+     * @param string $appSecret
+     */
+    public function useClientCredentials(string $appId, string $appSecret): void
+    {
+        $auth = new ClientCredentials($appId, $appSecret);
+
+        $this->auth = $auth;
+    }
+
+    /**
+     * @param string $clientId
+     * @param string $apiKey
+     */
+    public function useLegacyToken(string $clientId, string $apiKey): void
+    {
+        $auth = new LegacyCredentials($clientId, $apiKey);
+
+        $this->auth = $auth;
+    }
+
+    /**
+     * @param string $appId
+     * @param string $appSecret
+     * @param string $refreshToken
+     */
+    public function useRefreshToken(string $appId, string $appSecret, string $refreshToken): void
+    {
+        $auth = new RefreshCredentials($appId, $appSecret, $refreshToken);
+
+        $this->auth = $auth;
```

Comment:
Can we skip the `$auth` and just set it direct via `$this->auth = new Refresh...`?

Reply:
done